### PR TITLE
R doc 1777 map reduce output to collection

### PR DESCRIPTION
**Pull request by morhilai**

### CODE REVIEW

**Review by arekpalinski — COMMENTED** [review]
I think we're missing _very_ important detail here. The ability to define the _well known_ ID of the reference document (using the pattern) allows to _load_ it easily by another index and then load the actual artificial docs using `ReduceOutputs` property.

See here the example:

https://github.com/ravendb/ravendb/blob/b8163cb42d94f9a1799fee27454cd77e9bb4e30c/test/SlowTests/Server/Documents/Indexing/MapReduce/RavenDB_14418.cs#L237-L239

And Oren's blog about that:

https://ravendb.net/articles/ravendb-and-output-collection-pattern

File: Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown [hunk 1]
```diff
@@ -81,18 +81,37 @@ from 'Product/Sales'
...
+daily/monthly/yearly summaries very cheaply and easy.  
+
+In addition, you can also apply the usual operations on artificial documents (e.g. data 
+subscriptions or ETL).  
+
+Writing map-reduce outputs into documents allows you to define additional indexes on top of them 
+that give you the option to create recursive map-reduce operations. This way, you can do 
+daily/monthly/yearly summaries very cheaply and easy.  
+
+In addition, you can also apply the usual operations on documents (e.g. data subscriptions or ETL).
```
[the remark this answers is not in the document]
Comment (morhilai) [reply Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown:102]: Fixed

File: Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown [hunk 2]
```diff
@@ -89,98 +89,95 @@ called _"artificial",_ you need to define the target collection using the `Outpu
...
-- the current index is already working on (e.g. index on `DailyInvoices` collections outputs to `DailyInvoices`),
-- the current index is loading a document from it (e.g. index has `LoadDocument(id, "Invoices")` outputs to `Invoices`), 
-- it is processed by another map-reduce index that outputs results to a collection that the current index is working on (e.g. one index on `Invoices` collection outputs to `DailyInvoices`, another index on `DailyInvoices` outputs to `Invoices`)
+#### IDs
 
-Since that would result in the infinite indexing loop (the index puts an artificial document what triggers the indexing and so on), you will get the detailed error on attempt to create such invalid construction.
+The identifiers of map reduce output documents are generated as:  
 
-{WARNING/}
+`<OutputCollectionName>/<hash-of-reduce-key>`
```
Comment (arekpalinski) [inline Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown:121]: Actually there are two numbers there:

`<OutputCollectionName>/<numeric>/<hash-of-reduce-key>`

where `<numeric>` is an unique (cluster wide), incrementing number. Technically it is the Raft command id.

We need this to properly handle side by side indexing. See the comments in the related issue:

https://issues.hibernatingrhinos.com/issue/RavenDB-11488#focus=Comments-67-40981.0-0

File: Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown [hunk 3]
```diff
@@ -89,98 +89,95 @@ called _"artificial",_ you need to define the target collection using the `Outpu
...
+#### IDs
 
-Since that would result in the infinite indexing loop (the index puts an artificial document what triggers the indexing and so on), you will get the detailed error on attempt to create such invalid construction.
+The identifiers of map reduce output documents are generated as:  
 
-{WARNING/}
+`<OutputCollectionName>/<hash-of-reduce-key>`  
 
-{WARNING:Existing collection}
+The index in [example i](../indexes/map-reduce-indexes#example-i) might generate an output document ID like this:
```
Comment (arekpalinski) [inline Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown:123]:
```suggestion
The index in [Example I](../indexes/map-reduce-indexes#example-i) might generate an output document ID like this:  
```

File: Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown [hunk 4]
```diff
@@ -89,98 +89,95 @@ called _"artificial",_ you need to define the target collection using the `Outpu
...
 
-{WARNING:Existing collection}
+The index in [example i](../indexes/map-reduce-indexes#example-i) might generate an output document ID like this:  
 
-Creating a map-reduce index which defines the output collection that already exists and it contains documents will result in an error. You need to delete all documents
-from the relevant collection before creating the index or output the results to a different one.
+`DailyProductSales/35/14369232530304891504`  
 
-{WARNING/}
+The numeric part is the hash of the reduce key values, in this case: `hash(Product, Month)`.
```
Comment (arekpalinski) [inline Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown:127]: See comment above - there are two numeric parts.

File: Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown [hunk 5]
```diff
@@ -89,98 +89,95 @@ called _"artificial",_ you need to define the target collection using the `Outpu
...
-### Artificial Document IDs
+If the aggregation value for a given reduce key changes then we overwrite the artificial 
+document. It will get removed once there is no result for a given reduce key.  
 
-The identifiers of artificial documents are generated as:
+The identifiers of reference documents follow some pattern you choose, and this pattern 
+determines which output documents are held by a given reference document.  
 
-- `<OutputCollectionName>/<hash-of-reduce-key>`
+The index in [example i](../indexes/map-reduce-indexes#example-i) has this pattern for reference documents:
```
Comment (arekpalinski) [inline Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown:135]:
```suggestion
The index in [Example I](../indexes/map-reduce-indexes#example-i) has this pattern for reference documents:  
```

File: Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown [hunk 6]
```diff
@@ -89,98 +89,95 @@ called _"artificial",_ you need to define the target collection using the `Outpu
...
 The map-reduce output documents are configured with these properties of 
 `IndexDefinition`:  
 
-{CODE:csharp syntax@Indexes\MapReduceIndexes.cs /}
+{CODE:csharp syntax_0@Indexes\MapReduceIndexes.cs /}
+
+| Parameters | Type | Description |
+| - | - | - |
+| **OutputReduceToCollection** | `string` | Collection name for the output documents. |
+| **PatternReferencesCollectionName** | `string` | Collection name for the reference documents. |
```
Comment (arekpalinski) [inline Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown:163]: Maybe it's worth to mention that it's optional and the default collection name will be `<OutputReduceToCollection>/References`

File: Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown [hunk 7]
```diff
@@ -191,7 +188,7 @@ In the **LINQ** index example above (which inherits `AbstractIndexCreationTask`)
 the reference document ID pattern is set with a lambda expression:  
 
 {CODE-BLOCK:csharp}
-PatternForOutputReduceToCollectionReferences = x => $"sales/monthly/{x.Month}";
+PatternForOutputReduceToCollectionReferences = x => $"sales/daily/{x.Date:yyyy-MM-dd}";
```
Comment (arekpalinski) [inline Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown:191]: Maybe let's mention somewhere that the supported formatting like `yyyy-MM-dd` here is the same as `StringBuilder.AppendFormat` supports:

https://docs.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.appendformat?view=netcore-3.1

File: Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown (hunk 2 above)
Comment (morhilai) [reply Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown:121]: Right, wasn't sure if it was important to mention. Fixed ✔️

File: Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown (hunk 3 above)
Comment (morhilai) [reply Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown:123]: done

File: Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown (hunk 4 above)
Comment (morhilai) [reply Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown:127]: fixed

File: Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown (hunk 5 above)
Comment (morhilai) [reply Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown:135]: done

File: Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown (hunk 6 above)
Comment (morhilai) [reply Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown:163]: Done

File: Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown (hunk 7 above)
Comment (morhilai) [reply Documentation/4.2/Raven.Documentation.Pages/indexes/map-reduce-indexes.dotnet.markdown:191]: Okay, adding this:
>The pattern is built using the same syntax as [the `StringBuilder.AppendFormat` method](https://docs.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.appendformat). See [here](https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-date-and-time-format-strings) to learn about the date formatting in particular.

**Pull request merged**